File: 3.x.x/crates-utopia/domains/src/registrar/adapter.rs
use crate::cache::Cache;
use crate::error::DomainsError;

use super::{
    Contacts, NameserverUpdate, Price, RegistrarDomain, Renewal, SuggestItem, SuggestQuery,
    TransferStatus, UpdateDetails,
};

/// Registrar adapter (PHP `Utopia\Domains\Registrar\Adapter`).
pub trait Adapter: Send + Sync + std::fmt::Debug {
    /// Adapter identifier (`getName()`).
    fn get_name(&self) -> String;

    /// Check availability.
    fn available(&self, domain: &str) -> Result<bool, DomainsError>;

    /// Purchase a domain.
    #[allow(clippy::too_many_arguments)]
    fn purchase(
        &self,
        domain: &str,
        contacts: Contacts,
        period_years: i64,
        nameservers: Vec<String>,
        autorenew_enabled: bool,
        purchase_price: Option<f64>,
    ) -> Result<String, DomainsError>;

    /// Suggest domains.
    #[allow(clippy::too_many_arguments)]
    fn suggest(
        &self,
        query: SuggestQuery,
        tlds: Vec<String>,
        limit: Option<i64>,
        filter_type: Option<&str>,
        price_max: Option<i64>,
        price_min: Option<i64>,
    ) -> Result<std::collections::HashMap<String, SuggestItem>, DomainsError>;

    /// Supported TLDs.
    fn tlds(&self) -> Result<Vec<String>, DomainsError>;

    /// Fetch domain details.
    fn get_domain(&self, domain: &str) -> Result<RegistrarDomain, DomainsError>;

    /// Update domain details.
    fn update_domain(&self, domain: &str, details: &UpdateDetails) -> Result<bool, DomainsError>;

    /// Update nameservers. Default matches PHP (`Method not implemented`).
    fn update_nameservers(
        &self,
        _domain: &str,
        _nameservers: Vec<String>,
    ) -> Result<NameserverUpdate, DomainsError> {
        Err(DomainsError::generic("Method not implemented", 0))
    }

    /// Fetch a price.
    fn get_price(
        &self,
        domain: &str,
        period_years: i64,
        reg_type: &str,
        ttl: u64,
    ) -> Result<Price, DomainsError>;

    /// Renew a domain.
    fn renew(&self, domain: &str, period_years: i64) -> Result<Renewal, DomainsError>;

    /// Transfer a domain.
    fn transfer(
        &self,
        domain: &str,
        auth_code: &str,
        purchase_price: Option<f64>,
    ) -> Result<String, DomainsError>;

    /// Fetch an EPP auth code.
    fn get_auth_code(&self, domain: &str) -> Result<String, DomainsError>;

    /// Transfer status.
    fn check_transfer_status(&self, domain: &str) -> Result<TransferStatus, DomainsError>;

    /// Cancel pending purchases.
    fn cancel_purchase(&self) -> Result<bool, DomainsError>;

    /// Default nameservers used when purchase is called with an empty list.
    fn set_default_nameservers(&mut self, nameservers: Vec<String>);

    /// Attach a cache (used by `getPrice`).
    fn set_cache(&mut self, cache: Option<Cache>);

    /// TCP connect timeout in seconds.
    fn set_connect_timeout(&mut self, connect_timeout: u64);

    /// Request timeout in seconds.
    fn set_timeout(&mut self, timeout: u64);
}

/// Shared HTTP adapter state (timeouts, cache, nameservers).
#[derive(Debug)]
pub(crate) struct AdapterState {
    pub default_nameservers: Vec<String>,
    pub cache: Option<Cache>,
    pub connect_timeout: u64,
    pub timeout: u64,
}

impl Default for AdapterState {
    fn default() -> Self {
        Self {
            default_nameservers: Vec::new(),
            cache: None,
            connect_timeout: 5,
            timeout: 10,
        }
    }
}

impl AdapterState {
    pub(crate) fn nameservers_or_default(&self, nameservers: Vec<String>) -> Vec<String> {
        if nameservers.is_empty() {
            self.default_nameservers.clone()
        } else {
            nameservers
        }
    }
}

/// PHP `empty($value)` for contact strings.
pub(crate) fn php_empty(value: &str) -> bool {
    value.is_empty() || value == "0"
}

/// Approximate PHP `FILTER_VALIDATE_EMAIL`.
pub(crate) fn is_valid_email(email: &str) -> bool {
    let Some((local, domain)) = email.rsplit_once('@') else {
        return false;
    };
    if local.is_empty() || domain.is_empty() || local.len() > 64 {
        return false;
    }
    if email.contains(' ') || email.contains('\n') || email.contains('\r') {
        return false;
    }
    domain.contains('.')
}

File: src/Domains/Registrar/Adapter.php
<?php

declare(strict_types=1);

namespace Utopia\Domains\Registrar;

use Psr\Http\Client\ClientInterface;
use Utopia\Client\Adapter\Curl\Client as CurlAdapter;
use Utopia\Client\Client;
use Utopia\Domains\Adapter as DomainsAdapter;
use Utopia\Domains\Cache;
use Utopia\Domains\Registrar;

abstract class Adapter extends DomainsAdapter
{
    /**
     * Default nameservers for domain registration
     */
    protected array $defaultNameservers = [];

    /**
     * Cache instance
     */
    protected ?Cache $cache = null;

    /**
     * Connection timeout in seconds
     */
    protected int $connectTimeout = 5;

    /**
     * Request timeout in seconds
     */
    protected int $timeout = 10;

    /**
     * An injected transport owns its timeout, TLS, redirect and connection settings.
     */
    protected ?ClientInterface $client = null;

    private ?Client $defaultClient = null;

    protected function getHttpClient(): ClientInterface
    {
        if ($this->client instanceof ClientInterface) {
            return $this->client;
        }

        $this->defaultClient ??= new Client(new CurlAdapter());

        return $this->defaultClient
            ->withConnectTimeout($this->connectTimeout)
            ->withTimeout($this->timeout)
            ->withSslVerification(true)
            ->withFollowRedirects(false);
    }

    /**
     * Set default nameservers
     */
    public function setDefaultNameservers(array $nameservers): void
    {
        $this->defaultNameservers = $nameservers;
    }

    /**
     * Set cache instance
     */
    public function setCache(?Cache $cache): void
    {
        $this->cache = $cache;
    }

    /**
     * Set connection timeout for the default transport. Injected clients are unchanged.
     */
    public function setConnectTimeout(int $connectTimeout): void
    {
        $this->connectTimeout = $connectTimeout;
    }

    /**
     * Set request timeout for the default transport. Injected clients are unchanged.
     */
    public function setTimeout(int $timeout): void
    {
        $this->timeout = $timeout;
    }

    /**
     * Get the name of the adapter
     */
    abstract public function getName(): string;

    /**
     * Check if domains are available
     *
     * @param array<string> $domains Domain names to check
     * @return array<string, bool> Availability keyed by domain name
     */
    abstract public function available(array $domains): array;

    /**
     * Purchase a domain
     *
     * @param  float|null  $purchasePrice Required if domain is premium
     * @return string Order ID
     */
    abstract public function purchase(string $domain, array|Contact $contacts, int $periodYears = 1, array $nameservers = [], bool $autorenewEnabled = false, ?float $purchasePrice = null): string;

    /**
     * Suggest domain names
     *
     * @param  array  $query
     * @param  string|null $filterType Filter results by type: 'premium', 'suggestion', or null for both
     */
    abstract public function suggest(array|string $query, array $tlds = [], ?int $limit = null, ?string $filterType = null, ?int $priceMax = null, ?int $priceMin = null): array;

    /**
     * Get the TLDs supported by the adapter
     */
    abstract public function tlds(): array;

    /**
     * Get the domain information
     */
    abstract public function getDomain(string $domain): Domain;

    /**
     * Update the domain information
     */
    abstract public function updateDomain(string $domain, UpdateDetails $details): bool;

    /**
     * Update the nameservers for a domain
     *
     * @throws \Exception
     */
    public function updateNameservers(string $domain, array $nameservers): array
    {
        throw new \Exception('Method not implemented');
    }

    /**
     * Get the price of a domain
     */
    abstract public function getPrice(string $domain, int $periodYears = 1, string $regType = Registrar::REG_TYPE_NEW, int $ttl = 3600): Price;

    /**
     * Renew a domain
     */
    abstract public function renew(string $domain, int $periodYears): Renewal;

    /**
     * Transfer a domain
     *
     * @param  float|null  $purchasePrice Required if domain is premium
     * @return string Order ID
     */
    abstract public function transfer(string $domain, string $authCode, ?float $purchasePrice = null): string;

    /**
     * Get the authorization code for an EPP domain
     */
    abstract public function getAuthCode(string $domain): string;

    /**
     * Check transfer status for a domain
     */
    abstract public function checkTransferStatus(string $domain): TransferStatus;

    /**
     * Cancel pending purchase orders
     */
    abstract public function cancelPurchase(): bool;
}
